--- extensions/naxsi/naxsi_src/naxsi_utils.c.orig
+++ extensions/naxsi/naxsi_src/naxsi_utils.c
@@ -378,11 +378,11 @@
                   "whitelist targets |NAME");


-      strncat(*fullname, (const char *) "#", 1);
+      strncat(*fullname, (const char *) "#", 2);
     }
     strncat(*fullname, (const char *) custloc_array(curr->br->custom_locations->elts)[uri_idx].target.data,
            custloc_array(curr->br->custom_locations->elts)[uri_idx].target.len);
-    strncat(*fullname, (const char *) "#", 1);
+    strncat(*fullname, (const char *) "#", 2);
     strncat(*fullname, (const char *) custloc_array(curr->br->custom_locations->elts)[name_idx].target.data,
            custloc_array(curr->br->custom_locations->elts)[name_idx].target.len);
   }
@@ -400,7 +400,7 @@
                   "whitelist targets |NAME");


-      strncat(*fullname, (const char *) "#", 1);
+      strncat(*fullname, (const char *) "#", 2);
     }


@@ -415,7 +415,7 @@

     *fullname = ngx_pcalloc(cf->pool, custloc_array(curr->br->custom_locations->elts)[name_idx].target.len + 2);
     if (curr->br->target_name)
-      strncat(*fullname, (const char *) "#", 1);
+      strncat(*fullname, (const char *) "#", 2);
     strncat(*fullname, (const char *) custloc_array(curr->br->custom_locations->elts)[name_idx].target.data,
            custloc_array(curr->br->custom_locations->elts)[name_idx].target.len);
   }
--- ./extensions/naxsi/naxsi_src/naxsi_runtime.c.orig
+++ ./extensions/naxsi/naxsi_src/naxsi_runtime.c
@@ -725,10 +725,10 @@
   tmp_hashname.data = ngx_pcalloc(req->pool, tmp_hashname.len+2);
   if (target_name) {
     tmp_hashname.len++;
-    strncat((char*)tmp_hashname.data, "#", 1);
+    strncat((char*)tmp_hashname.data, "#", 2);
   }
   strncat((char*) tmp_hashname.data, (char*)req->uri.data, req->uri.len);
-  strncat((char*)tmp_hashname.data, "#", 1);
+  strncat((char*)tmp_hashname.data, "#", 2);
   strncat((char*)tmp_hashname.data, (char*)name->data, name->len);

   NX_DEBUG(_debug_whitelist_compat, NGX_LOG_DEBUG_HTTP, req->connection->log, 0, "hashing MIX [%V] ($URL:x|$X_VAR:y) or ($URL:x|$X_VAR:y|NAME)", &tmp_hashname);
--- ./extensions/naxsi/naxsi_src/naxsi_skeleton.c.orig
+++ ./extensions/naxsi/naxsi_src/naxsi_skeleton.c
@@ -414,8 +414,8 @@
   libjct_sql->sc_tag->data = ngx_pcalloc(cf->pool, 18 /* LIBINJECTION_SQL */);
   libjct_xss->sc_tag->data = ngx_pcalloc(cf->pool, 18 /* LIBINJECTION_XSS */);
   if (!libjct_sql->sc_tag->data || !libjct_xss->sc_tag->data) return (NGX_ERROR); /* LCOV_EXCL_LINE */
-  strncpy((char *)libjct_sql->sc_tag->data, (char *)"$LIBINJECTION_SQL", 17);
-  strncpy((char *)libjct_xss->sc_tag->data, (char *)"$LIBINJECTION_XSS", 17);
+  strncpy((char *)libjct_sql->sc_tag->data, (char *)"$LIBINJECTION_SQL", 18);
+  strncpy((char *)libjct_xss->sc_tag->data, (char *)"$LIBINJECTION_XSS", 18);
   libjct_xss->sc_tag->len = 17;
   libjct_sql->sc_tag->len = 17;
   libjct_sql->sc_score = 8;

